control_panel_api/management/commands/migrate_lambdas_data_2_users3buckets.py

```diff
@@ -0,0 +1,108 @@
+    """
+    NOTE: Needs permission to perform `iam:ListAttachedRolePolicies` action
+    """
+
+
+    help = ("Add UserS3Bucket records to reflect access to team S3 buckets "
+            "granted by AWS Lambda functions.")
+
+    def handle(self, *args, **options):
+        iam = boto3.client('iam')
+
+        users = User.objects.all()
+        for user in users:
+            role_name = user.iam_role_name
+
+            logger.info(
+                f'Migrating user "{user.username}" ("{user.auth0_id}"): '
+                f'Reading policies from role "{role_name}"...'
+            )
+            try:
+                policies = iam.list_attached_role_policies(
+                    RoleName=role_name,
+                    MaxItems=1000,
+                )
+            except ClientError as e:
+                if e.response['Error']['Code'] in ('NoSuchEntity', 'ValidationError'):
+                    logger.warning(
+                        f'Error reading policies from IAM role "{role_name}": '
+                        f'user "{user.username}" ("{user.pk}"): {e}'
+                    )
+                    continue
+                else:
+                    raise e
+
+            if policies:
+                for policy in policies["AttachedPolicies"]:
+                    policy_name = policy["PolicyName"]
+                    if _is_eligible(policy_name):
+                        bucket_name = _bucket_name(policy_name)
+                        s3bucket = S3Bucket.objects.filter(name=bucket_name)
+                        if not s3bucket.exists():
+                            logger.warning(
+                                f'S3 bucket "{bucket_name}" not found: '
+                                f'corresponding to IAM policy "{policy_name}"'
+                            )
+                            continue
+
+                        s3bucket = s3bucket.first()
+
+                        users3bucket = UserS3Bucket.objects.filter(
+                            user=user,
+                            s3bucket=s3bucket,
+                        )
+                        if not users3bucket.exists():
+                            try:
+                                if not DRYRUN:
+                                    UserS3Bucket.objects.create(
+                                        user=user,
+                                        s3bucket=s3bucket,
+                                        access_level=READWRITE,
+                                    )
+                                logger.info(
+                                    f'UserS3Bucket created: '
+                                    f'({user.username}, {bucket_name})'
+                                )
+                            except Exception as e:
+                                logger.critical(
+                                    f'Failed to create UserS3Bucket: '
+                                    f'for ({user.username}, {bucket_name}): {e}'
```

Comment:
`logger.critical` but it doesn't break out of the loop or stop the migration?

Reply:
I see what you mean but I still think it would be annoying and draconian to stop the migration halfway through in production just because we can't create a single record for example.

The idea was to output this as critical to highlight the problem and investigate/fix.

This is probably "never" going to happen but I wanted to handle anyway to be defensive. Although I just realised that I swallow the exception here, and that's bad as I wouldn't know what went wrong in the unlikely event there is an exception. I will change the message to display the error in the message.

Reply:
Oh actually, I didn't swallow the exception error, it was just the weird GH diff formatting 🎉